Add support for user data

Trey2k opened the pull request:
Hello, I was looking around for lightweight C json parsers and I found this project. It looks great and I love the idea of using callbacks for key and values. As well as the fact that you don't try to parse the data type for us it seems.

One thing I think could improve this is support for user data. This way static globals are not required to parse information. So that is what this PR adds.

saadshams commented:
Hello, Thanks for your comments. There's a bit of rebranding going on due to back and forth formalities with vcpkg and you will see some changes.

With the callback and the data type conversions, I wanted to put the end-user in the driving seat since they are aware of the schemas and the contracts, while keeping the API clean and simple, and let the library do one thing and one thing only: parse the key-value pairs or values in the case of JSON arrays.

I'm not understanding the idea behind the user data; it seems like you want to pass some kind of token or intend to pass allocations.

saadshams commented:
I have merged the changes; just so you know, I have also incorporated your changes into the rebranded repo. Once it goes through vcpkg, I'll then be able to publish the new repo formally.

Trey2k commented:
> To get the complete idea behind the userData, it seems like you want to pass some kind of token or allocations, and the callback populates properties with values on the allocations. Is that the correct assumption? Any examples just to be clear?

I have rewrote the first example to utilize the strategy i had in mind for this feature.

```c
#include "parse.h"
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char *name;
    int age;
} Person;

static void callback(const char *const key, const char *const value, const char *const parentKey, void* userData) {
    Person *person = (Person *) userData;
    if (strcmp(key, "name") == 0) person->name = strdup(value);
    if (strcmp(key, "age") == 0) person->age = atoi(value);
}

int main(void) {
    Person person;

    char *json = "{\"name\": \"John Doe\", \"age\": 25}";
    nanojson_parse_object(json, callback, NULL, &person);

    assert(strcmp(person.name, "John Doe") == 0);
    assert(person.age == 25);

    printf("Name: %s, Age: %d", person.name, person.age);
    free(person.name);
    return 0;
}
```
    
> I have merged the changes; just so you know, I have also incorporated your changes into the rebranded repo. Once it goes through vcpkg, I'll then be able to publish the new repo formally.
    
I am not familiar with vcpkg at all, do you mean to say the github repo is self is changing?

Trey2k commented:
> For future updates and pull requests, please lookup saadshams/nanojsonc.

Got you that makes sense, It might be too late now, but github has a feature that allows you to rename the repo. When you do so both the old repo name and the new one will work. It will redirect you.

saadshams commented:
I see. I didn't know about the redirect feature. Refactoring an existing repo felt quite chaotic due to several references in vcpkg-related CMake and port files since integrations are very tight. I'm still waiting on the outcome of the submission and some more work to make it easier.

Are you good with CMake and CTest? Tests are running but not when `ctest --submit`, any idea?

Trey2k commented:
> Are you good with CMake and CTest? Tests are running but not when ctest --submit, any idea?

Sadly I am only somewhat familiar with CMake, I normally work with the SCons build system. I potentially can take a look though when I get some time, but I dont imagine I have much to bring to the table for CMake.
